sycl/include/sycl/detail/vector_arith.hpp

```diff
@@ -73,19 +73,27 @@ struct VecOperators {
         is_logical, vec<fixed_width_signed<sizeof(element_type)>, N>, Self>;
 
     BinOp Op{};
-    if constexpr (is_host || N == 1 ||
-                  std::is_same_v<element_type, ext::oneapi::bfloat16>) {
-      result_t res{};
-      for (size_t i = 0; i < N; ++i)
-        if constexpr (is_logical)
-          res[i] = Op(Args[i]...) ? -1 : 0;
-        else
-          res[i] = Op(Args[i]...);
-      return res;
-    } else {
-      using vector_t = typename Self::vector_t;
-
-      auto res = [&](auto... xs) {
+#ifdef __has_extension
+#if __has_extension(attribute_ext_vector_type)
+    // ext_vector_type's bool vectors are mapped onto <N x i1> and have
+    // different memory layout than sycl::vec<bool ,N> (which has 1 byte per
+    // element). As such we perform operation on int8_t and then need to
+    // create bit pattern that can be bit-casted back to the original
+    // sycl::vec<bool, N>. This is a hack actually, but we've been doing
+    // that for a long time using sycl::vec::vector_t type.
+    using vec_elem_ty =
+        typename detail::map_type<element_type, bool, /*->*/ std::uint8_t,
+#if (!defined(_HAS_STD_BYTE) || _HAS_STD_BYTE != 0)
+                                  std::byte, /*->*/ std::int8_t,
+#endif
+#ifdef __SYCL_DEVICE_ONLY__
+                                  half, /*->*/ _Float16,
+#endif
+                                  element_type, /*->*/ element_type>::type;
+    if constexpr (N != 1 && !check_type_in_v<vec_elem_ty, bool, half,
+                                             ext::oneapi::bfloat16>) {
+      using vec_t = vec_elem_ty __attribute__((ext_vector_type(N)));
+      auto tmp = [&](auto... xs) {
         // Workaround for https://github.com/llvm/llvm-project/issues/119617.
         if constexpr (sizeof...(Args) == 2) {
```

Comment:
When would this not be the case? `Op` has type `BinOp`, which suggests it's a binary operator.

Reply:
Renamed the type.